File: src/types/task.ts
/**
 * タスク実行アクションの定数定義
 */
export const TASK_ACTION_TYPES = [
    "AI_ANALYZE",
    "AUTOMATED_REMEDIATE",
    "SYSTEM_NOTIFICATION",
    "EXTERNAL_WEBHOOK",
    "KILL_SWITCH",
    "ESCALATE",
] as const;

export type TaskActionType = (typeof TASK_ACTION_TYPES)[number];

/**
 * タスク優先度 (1: 最高/即時 - 5: 最低/バッチ)
 */
export type TaskPriority = 1 | 2 | 3 | 4 | 5;

/**
 * タスク実行レベル
 * AUTO: 自動実行
 * SEMI_AUTO: 確認後自動実行
 * MANUAL: 人間が実行
 * MONITOR: 監視のみ
 */
export type TaskExecutionLevel = "AUTO" | "SEMI_AUTO" | "MANUAL" | "MONITOR";

/**
 * タスクルール定義
 * EventDetectorが検知したイベント名に紐づくルール
 */
export interface TaskRule {
    ruleId: string;
    eventName: string;
    severity: TaskSeverity;
    actionType: TaskActionType;
    executionLevel: TaskExecutionLevel;
    priority: TaskPriority;
    description: string;
    executionParams: {
        targetEndpoint?: string;
        scriptIdentifier?: string;
        notificationChannel?: string;
        promptTemplate?: string;
    };
    guardrails: {
        requireHumanApproval: boolean;
        timeoutMs: number;
        maxRetries: number;
    };
}

/**
 * 重大度分類
 */
export const TASK_SEVERITIES = [
    "CRITICAL",
    "HIGH",
    "MEDIUM",
    "LOW",
    "INFO",
] as const;

export type TaskSeverity = (typeof TASK_SEVERITIES)[number];

/**
 * 生成されたタスク
 */
export interface GeneratedTask {
    taskId: string;
    ruleId: string;
    eventName: string;
    severity: TaskSeverity;
    actionType: TaskActionType;
    executionLevel: TaskExecutionLevel;
    priority: TaskPriority;
    description: string;
    executionParams: TaskRule["executionParams"];
    guardrails: TaskRule["guardrails"];
    sourceLog: {
        traceId: string;
        message: string;
        boundary: string;
        level: number;
        timestamp: string;
    };
    createdAt: string;
}

/**
 * タスク実行結果
 */
export type TaskDispatchStatus = "dispatched" | "blocked_approval" | "skipped" | "failed";

export interface TaskResult {
    taskId: string;
    ruleId: string;
    status: TaskDispatchStatus;
    dispatchedAt: string;
    error?: string;
}

File: packages/server/internal/domain/task.go
package domain

import "time"

// TaskActionType タスクアクション種別
type TaskActionType string

const (
	ActionAIAnalyze          TaskActionType = "AI_ANALYZE"
	ActionAutomatedRemediate TaskActionType = "AUTOMATED_REMEDIATE"
	ActionSystemNotification TaskActionType = "SYSTEM_NOTIFICATION"
	ActionExternalWebhook    TaskActionType = "EXTERNAL_WEBHOOK"
	ActionKillSwitch         TaskActionType = "KILL_SWITCH"
	ActionEscalate           TaskActionType = "ESCALATE"
)

// TaskSeverity 重大度
type TaskSeverity string

const (
	SeverityCritical TaskSeverity = "CRITICAL"
	SeverityHigh     TaskSeverity = "HIGH"
	SeverityMedium   TaskSeverity = "MEDIUM"
	SeverityLow      TaskSeverity = "LOW"
	SeverityInfo     TaskSeverity = "INFO"
)

// TaskExecutionLevel 実行レベル
type TaskExecutionLevel string

const (
	ExecLevelAuto     TaskExecutionLevel = "AUTO"
	ExecLevelSemiAuto TaskExecutionLevel = "SEMI_AUTO"
	ExecLevelManual   TaskExecutionLevel = "MANUAL"
	ExecLevelMonitor  TaskExecutionLevel = "MONITOR"
)

// TaskPriority 優先度 (1=最高, 5=最低)
type TaskPriority int

// TaskDispatchStatus ディスパッチ結果
type TaskDispatchStatus string

const (
	StatusPending         TaskDispatchStatus = "pending"
	StatusDispatched      TaskDispatchStatus = "dispatched"
	StatusBlockedApproval TaskDispatchStatus = "blocked_approval"
	StatusApproved        TaskDispatchStatus = "approved"
	StatusRejected        TaskDispatchStatus = "rejected"
	StatusCompleted       TaskDispatchStatus = "completed"
	StatusSkipped         TaskDispatchStatus = "skipped"
	StatusFailed          TaskDispatchStatus = "failed"
)

// TaskRule タスク生成ルール
type TaskRule struct {
	RuleID         string
	EventName      string
	Severity       TaskSeverity
	ActionType     TaskActionType
	ExecutionLevel TaskExecutionLevel
	Priority       TaskPriority
	Description    string
	ExecParams     ExecParams
	Guardrails     Guardrails
}

// ExecParams 実行パラメータ
type ExecParams struct {
	TargetEndpoint      string
	ScriptIdentifier    string
	NotificationChannel string
	PromptTemplate      string
}

// Guardrails ガードレール
type Guardrails struct {
	RequireHumanApproval bool
	TimeoutMs            int
	MaxRetries           int
}

// GeneratedTask 生成されたタスク
type GeneratedTask struct {
	TaskID         string
	RuleID         string
	EventName      string
	Severity       TaskSeverity
	ActionType     TaskActionType
	ExecutionLevel TaskExecutionLevel
	Priority       TaskPriority
	Description    string
	ExecParams     ExecParams
	Guardrails     Guardrails
	SourceLog      SourceLogInfo
	CreatedAt      time.Time
}

// SourceLogInfo タスクの元になったログ情報
type SourceLogInfo struct {
	TraceID   string
	Message   string
	Boundary  string
	Level     LogLevel
	Timestamp time.Time
}

// TaskResult タスク実行結果
type TaskResult struct {
	TaskID       string
	RuleID       string
	Status       TaskDispatchStatus
	DispatchedAt time.Time
	Error        string
}

// StoredTask はDB永続化されたタスク（ステータス管理付き）
type StoredTask struct {
	GeneratedTask
	Status       TaskDispatchStatus
	ErrorMessage string
	UpdatedAt    time.Time
}

// SeverityOrder 重大度の比較用序列
var SeverityOrder = map[TaskSeverity]int{
	SeverityInfo:     0,
	SeverityLow:      1,
	SeverityMedium:   2,
	SeverityHigh:     3,
	SeverityCritical: 4,
}

// SeverityGTE severity a >= b か判定
func SeverityGTE(actual, threshold TaskSeverity) bool {
	return SeverityOrder[actual] >= SeverityOrder[threshold]
}
